follow_topics: Add support to follow a topic from the message header and recent topics UI.

prakhar1144 opened the pull request:
- [x]  Add support to follow a topic from the message header bar. 
- [x] Add support to follow a topic via recent conversations UI.

<!-- Describe your pull request here.-->

Fixes #25915.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

* [All Tooltips screenshots](https://github.com/zulip/zulip/pull/26353#discussion_r1292448407)

  <details><summary>Popover - 1</summary>
  <img src="https://github.com/zulip/zulip/assets/56781761/80a57adf-ef35-4a6c-8bb1-71af0bc9e79a">
  </img>
  </details> 

  <details><summary>Popover - 2</summary>
  <img src="https://github.com/zulip/zulip/assets/56781761/8ecd5706-9395-47b4-a00d-76575980f8c7">
  </img>
  </details>  

---


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Responsiveness and internationalization.
- [x] Strings and tooltips.
- [x] End-to-end functionality of buttons, interactions and flows.
- [x] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
Why is this one still marked as a draft? I see it has development environment guards that might make it possible to merge in advance of the rest of "follow topic", so wanted to check in on whether it's something I can try to merge.

prakhar1144 commented:
@timabbott 
This is ready as per the design alya suggested earlier.

But there is a pending design discussion for "What options should we show in muted/unmuted stream", we might have to make changes accordingly.
ref: https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1616605

timabbott commented on web/src/message_list_view.js:
```diff
@@ -210,6 +210,12 @@ function populate_group_from_message_container(group, message_container) {
+            group.stream_id,
+            group.topic,
+        );
+        group.all_visibility_policies = user_topics.all_visibility_policies;
```
I don't like adding `development` and `all_visibility_policies` to this object -- those are sorta global properties, and having a copy of them for each recipient group just feels weird.

Maybe it's hard to fix, and it's not exactly important, but it would be nice if we could have these properties just passed into template rendering.

timabbott replied:
(But I guess `all_visibility_policies` shouldn't be a thing that needs to in the DOM at all?)

timabbott replied:
I think an acceptable option would be to add a blank line followed by a comment above those two fields noting they are not specific to this group but that this is the easiest way we've figured out for passing the data to the template rendering.

prakhar1144 replied:
Added comment

timabbott commented on web/templates/recipient_row.hbs:
```diff
@@ -57,12 +57,27 @@
+                          role="button" tabindex="0" aria-haspopup="true" aria-label="{{t 'Mute or follow topic' }}"></i>
+                    {{else}}
+                        <i class="zulip-icon zulip-icon-mute-new recipient_bar_icon" data-tippy-content="{{t 'Change visibility policy'}}"
+                          role="button" tabindex="0" aria-haspopup="true" aria-label="{{t 'Change visibility policy' }}"></i>
```
A few notes. One is that this icon seems to be missing `cursor: pointer`. Next is that I think we want the icon to always show the current state -- for a topic in the default state, I think we want to use a "normal" icon, not a "mute" icon -- the latter should be reserved for if the topic is actually muted! 

And then finally, "visibility policy" is not a term intended to be user-facing. I think we want the tooltip to be something like a slightly fancier 2-line thing:

> You will get default notifications for this topic. (Or maybe in this case we just don't have a first line?)
> You have muted this topic.
> You have unmuted this topic.
> You follow this topic.

With a second like that's something like "Change notifications for this topic" as a click action description. I'm pretty sure we have an existing tooltip with this shape but cannot remember it. @amanagr do you recall?

amanagr replied:
We use one in `message_inline_image_tooltip`.

prakhar1144 replied:
* icon seems to be missing `cursor: pointer` -- Done
* I think we want the icon to always show the current state
  > Earlier we had planned to NOT show 'default' in the popover (which was ofcourse confusing). I have updated the behavior as per [our discussion in CZO](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1624043). 
  Also, we are still [discussing the icons](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20recipient.20bar/near/1617263) to use. For now, I have kept the icons for "Default" and "Unmute" as the same. The changes are not visible outside the dev env; so we can update the icons (once finalized) in a small commit. 
* I think we want the tooltip to be something like a slightly fancier 2-line thing:
  
  <details><summary>Default</summary>
  <img src="https://github.com/zulip/zulip/assets/56781761/63e7e511-aa24-4874-83ff-29520584e8ef">
  </img>
  </details> 
  
  <details><summary>Muted</summary>
  <img src="https://github.com/zulip/zulip/assets/56781761/f9bdcda9-336f-486c-b5a7-c353e76cfaee">
  </img>
  </details> 
  
  <details><summary>Unmuted</summary>
  <img src="https://github.com/zulip/zulip/assets/56781761/6ab43a8e-7a13-496f-80a0-650d1607633f">
  </img>
  </details> 
  
  <details><summary>Followed</summary>
  <img src="https://github.com/zulip/zulip/assets/56781761/d559a6e2-d8d4-4a36-a457-927de4c45a70">
  </img>
  </details>

prakhar1144 replied:
Should we use "Click to change the state of this topic", because these visibility policies affect things other (visually) than notifications. 

We use the "state" term in 'SETTINGS/TOPICS' panel.

timabbott replied:
I don't think "state" feels right -- let's discuss this detail in #design; I think this is a workable prototype but we'll probably refine it a bit more as we go.

timabbott commented on web/styles/app_components.css:
```diff
@@ -1089,3 +1089,11 @@ div.overlay {
+    i {
+        font-size: 17px;
+    }
+}
```
Hmm, I think this belongs in `message_row.css` or something that contains the recipient elements?

prakhar1144 replied:
Moved it to `zulip.css`. 

`.recipient_bar_controls`, `.recipient_row_date` etc lies there.

timabbott replied:
OK. We may move them out but they can go together, so this works.

timabbott commented on web/src/popover_menus.js:
```diff
@@ -633,6 +635,64 @@ export function initialize() {
+                    topic_name,
+                    visibility_policy,
+                );
+                instance.hide();
```
Just noting that this doesn't have a good way to do error handling -- it's probably OK as I don't think it will come up often enough to matter, but let's add a `// TODO: Figure out a good way to offer feedback if this request fails`.

prakhar1144 replied:
Comment added

timabbott commented:
> The current visibility_policy of the topic is highlighted.

I didn't notice it highlighted in the popover at least in the dark theme -- am I missing something or is it just super subtle?

timabbott commented:
Overall this looks pretty good; I posted a few comments and think we should plan to resolve those comments quickly and then merge this -- I feel pretty good about our being able to iterate with tweaks on top of this towards having a completed feature.

prakhar1144 commented:
> > The current visibility_policy of the topic is highlighted.
> 
> I didn't notice it highlighted in the popover at least in the dark theme -- am I missing something or is it just super subtle?

I should have written:
The current visibility_policy of the topic is highlighted **in the popover.**

prakhar1144 commented:
Updates in the recent push:
* Tweaks as per our [discussion in #design > Follow Topics UI](https://chat.zulip.org/#narrow/stream/101-design/topic/Follow.20Topics.20UI/near/1624043)
  * **In muted stream**: 4 options (Mute, Regular, Unmute, Follow) 
  * **In not muted stream**: 3 options (Mute, Regular, Follow) + Unmute is available only when the visibility_policy is set to 'Unmuted'.
  
 * Addressed the review comments (Please take a look at the respective comments; I have replied there)
 * Added second commit: Add support to follow topic via recent conversations. -- This is a small change (~50 lines) & closely related to what we are working on; so included in this PR.
 
    <details><summary>Mute - Unmute - Follow</summary>
    <img src="https://github.com/zulip/zulip/assets/56781761/ea031b7b-3d79-4e9d-8b19-9e604cb5000e">
    </img>
    </details> 
    
    <details><summary>Tooltip</summary>
    <img src="https://github.com/zulip/zulip/assets/56781761/73e036df-7ff8-4d7d-8c8d-e3be6082fbb7">
    </img>
    </details> 
    
  * [All tooltips screenshots](https://github.com/zulip/zulip/pull/26353#discussion_r1292448407)
  
  ---
  Note: we are still [discussing the icons](https://chat.zulip.org/#narrow/stream/101-design/topic/UI.20redesign.3A.20recipient.20bar/near/1617263) to use. For now, I have kept the icons for "Default" and "Unmute" as the same. The changes are not visible outside the dev env; so we can update the icons (once finalized) in a small commit.

prakhar1144 commented:
Hmm, So #25915 includes two things:
* Add support to follow topic from message header.
* Add support to follow topic via recent conversations UI.

I have fixed those in two separate commits each including the last para as:
`Fixes part of #25915` -- but zulipbot gives warning.

I think we should ignore this warning and close the issue manually. Ideally, the issue should have been two separate issues.

timabbott commented:
What Zulipbot is fussing about is the PR have this at the top but not the same in the commit message; often we'll just put that in the second commit if it's the last checkbox on an issue ... but I kinda agree the whole issue of mapping to checkboxes within an issue is annoying.

> Fixes #25915.

timabbott commented:
This feels great, merged, thanks for all the work on this @prakhar1144! 

Let's discuss in chat.zulip.org what else we want to do before turning this on there as well. (I think one option might be to make those `development` variables `true`)
